README.md
INSTALLING
================

Make sure [node.js](http://nodejs.org/#download) is installed. Then install `nodeload`:

1. Using [npm](http://npmjs.org/):

        curl http://npmjs.org/install.sh | sh       # install npm if not already installed
        npm install nodeload

2. From source:

        git clone git://github.com/benschmaus/nodeload.git
        cd nodeload
        npm link    # optional. enables require('nodeload/<module>') instead of require('./lib/<module>').

3. Or as a single file (this does not install the `nl.js` tool):

        wget https://github.com/benschmaus/nodeload/raw/master/nodeload.js

NODELOAD
================

`nodeload` is collection of independent [node.js](http://nodejs.org/) modules for load testing HTTP services.

As a developer, you should be able write load tests and get informative reports without having to learn another framework. You should be able to build by example and selectively use the parts of a tool that fit your task. Being a library means that you can use as much or as little of `nodeload` as makes sense, and you can create load tests with the power of a full programming language. For example, if you need to execute some function at a given rate, just use the [`'nodeload/loop'`](https://github.com/benschmaus/nodeload/tree/master/doc/loop.md) module, and write the rest yourself.

In addition, `nodeload` is built for operability. It can always be deployed by simply copying the single file, `nodeload.js`.

Look for examples in the [`examples/`](https://github.com/benschmaus/nodeload/tree/master/examples) directory and in test cases prefixed with "example" in [`test/`](https://github.com/benschmaus/nodeload/tree/master/test). Here are simple examples of each module:

### [nl](https://github.com/benschmaus/nodeload/tree/master/doc/nl.md)

`nl` is an [Apache Bench (ab)](http://httpd.apache.org/docs/2.0/programs/ab.html) like command line tool for running tests quickly. See the [nl documentation](https://github.com/benschmaus/nodeload/tree/master/doc/nl.md) for details.

    $ examples/test-server.js &    # starts a simple server on port 9000 to load test
    $ ./nl.js -c 10 -n 10000 -i 2 localhost:9000

will send 10,000 queries to http://localhost:9000 using 10 connections. Statistics are printed to the console and graphs can be seen at <http://localhost:8000/>.

### [nodeload](https://github.com/benschmaus/nodeload/tree/master/doc/nodeload.md)

The `nodeload` module is the primary interface for creating load tests. It includes all the other modules described below, so if you `require('nodeload')`, you don't need to `require()` any of the other ones. Look at the examples in [`examples/loadtesting.ex.js`](https://github.com/benschmaus/nodeload/tree/master/examples/loadtesting.ex.js) and [`examples/riaktest.ex.js`](https://github.com/benschmaus/nodeload/tree/master/examples/riaktest.ex.js) or read the [nodeload module documentation](https://github.com/benschmaus/nodeload/tree/master/doc/nodeload.md).

    var nl = require('nodeload');
    var loadtest = nl.run({
        host: 'localhost',
        port: 9000,
        timeLimit: 60,
        targetRps: 500,
        requestGenerator: function(client) {
            var request = client.request('GET', "/" + Math.floor(Math.random()*10000));
            request.end();
            return request;
        }
    });
    loadtest.on('end', function() { console.log('Load test done.'); });

### [remote](https://github.com/benschmaus/nodeload/tree/master/doc/remote.md)

The `remote` module provides a mechanism for running a distributed load test. See [`examples/remotetesting.ex.js`](https://github.com/benschmaus/nodeload/tree/master/examples/remotetesting.ex.js) and [`examples/remote.ex.js`](https://github.com/benschmaus/nodeload/tree/master/examples/remote.ex.js) for examples or read the [remote module documentation](https://github.com/benschmaus/nodeload/tree/master/doc/remote.md).

Start slave instances:

    $ HTTP_PORT=10001 ./nodeload.js  # start a local slave instance on :10001
    $ HTTP_PORT=10002 ./nodeload.js  # start a 2nd slave instance on :10002
    
Create the distributed load test:

    var nl = require('nodeload/remote');
    var cluster = new nl.LoadTestCluster('localhost:8000', ['localhost:8002', 'localhost:8001']);
    cluster.run({
        host: 'localhost',
        port: 9000,
        timeLimit: 60,
        targetRps: 500,
        requestGenerator: function(client) {
            var request = client.request('GET', "/" + Math.floor(Math.random()*10000));
            request.end();
            return request;
        }
    });
    cluster.on('end', function() { console.log('Load test done.'); });

### [stats](https://github.com/benschmaus/nodeload/tree/master/doc/stats.md)

The `stats` module provides implementations of various statistics objects, like Histograms and Accumulators, and functions, like randomString(), and nextGaussian(). See the [stats module documentation](https://github.com/benschmaus/nodeload/tree/master/doc/stats.md).

    var stats = require('nodeload/stats');
    var histogram = new stats.Histogram();
    for (var i = 0; i < 1000; i++) 
        histogram.put(Math.abs(Math.floor(stats.nextGaussian())));
    console.log('Mean: ' + histogram.mean() + ', 99%: ' + histogram.percentile(0.99));

will output "`Mean: 0.852, 99%: 3`".

### [monitoring](https://github.com/benschmaus/nodeload/tree/master/doc/monitoring.md)

The `monitoring` module provides a way to track runtime statistics for code that is run concurrently. See [`test/monitoring.test.js`](https://github.com/benschmaus/nodeload/tree/master/test/monitoring.test.js) for examples or read the [monitoring module documentation](https://github.com/benschmaus/nodeload/tree/master/doc/monitoring.md).

    var monitoring = require('nodeload/monitoring');
    var monitor = new monitoring.Monitor('runtime');
    function asyncFunction() {
        var m = monitor.start();
        setTimeout(function() { m.end(); }, Math.floor(Math.random()*1000));
    }
    for (var i = 0; i < 1000; i++) { asyncFunction(); }
    process.on('exit', function() {
        console.log('Median runtime (ms): ' + monitor.stats['runtime'].percentile(0.5));
    });

will output "`Median runtime (ms): 497`".

### [reporting](https://github.com/benschmaus/nodeload/tree/master/doc/reporting.md)

The `reporting` module provides a way to graph values over time and present it in a auto-updating HTML page. See [`test/reporting.test.js`](https://github.com/benschmaus/nodeload/tree/master/test/reporting.test.js) for examples or read the [reporting module documentation](https://github.com/benschmaus/nodeload/tree/master/doc/reporting.md).

    var reporting = require('nodeload/reporting'), 
        stats = require('nodeload/stats'),
        report = reporting.REPORT_MANAGER.addReport('Random Numbers'),
        chart = report.getChart('Gaussian / Pareto vs. Time (minutes)');
    for (var timeout = 0; timeout < 5000; timeout+=500) {
        setTimeout(function() {
                chart.put({
                    'Pareto': stats.nextPareto(0, 100),
                    'Gaussian': stats.nextGaussian()
                });
            }, timeout);
    }

will display a graph on http://localhost:8000/ and save it to an HTML file in the local directory.

### [loop](https://github.com/benschmaus/nodeload/tree/master/doc/loop.md)

The `loop` module provides a way to execute a function at a set rate and concurrency. See [`test/loop.test.js`](https://github.com/benschmaus/nodeload/tree/master/test/loop.test.js) for examples or read the [loop module documentation](https://github.com/benschmaus/nodeload/tree/master/doc/loop.md) for details.

    var http = require('http'),
        loop = require('nodeload/loop'),
        requests = 0,
        client = http.createClient(80, 'www.google.com'),
        l = new loop.MultiLoop({
            fun: function(finished) { 
                client.request('GET', '/').end();
                requests++;
                finished();
            },
            rps: 10,
            duration: 3,
            concurrency: 5
        }).start();
    l.on('end', function() { console.log('Total requests: ' + requests) });

will output "`Total requests: 30`".

### [http](https://github.com/benschmaus/nodeload/tree/master/doc/http.md)

The `http` module provides a generic HTTP server that serves static files and that can be configured with new routes. See [`test/http.test.js`](https://github.com/benschmaus/nodeload/tree/master/test/http.test.js) for examples or read the [http module documentation](https://github.com/benschmaus/nodeload/tree/master/doc/http.md).

    var http = require('nodeload/http');
    var server = new http.HttpServer().start(10000);
    server.addRoute('^/hello$', function(url, req, res) {
        res.writeHead(200);
        res.end("Hello");
    });

<http://localhost:8000/package.json> will output the contents of `./package.json`, and <http://localhost:10000/resource> will display "Hello".


CONTRIBUTING
================
File bugs on [github](https://github.com/benschmaus/nodeload/issues), email any of the authors, and fork away. [doc/developers.md](https://github.com/benschmaus/nodeload/tree/master/doc/developers.md) has brief instructions on getting tests up and running, and will hold more design details in the future. Contributions are always welcome.

RELEASE-NOTES.md
## v0.4.0 (HEAD) ##

Compatible with node v0.4.x

Features:

* Add graphJmx() and graphProcess(); deprecate spawnAndMonitor(). These provide an easy way to graph JMX attributes as well as output from external processes, such as iostat.
* rps is a separate stat from result codes
* Test Results page timestamp does not auto-update on open
* X-axes now use real timestamps rather than minutes since test start

## v0.3.0 (2011/06/16) ##

Compatible with node v0.3.x

Features:

* Add /console/console.html, a jQuery based UI for connecting to multiple nodeload instances simultaneously
* jmxstat/jmxstat.jar allows command line polling of JMX attributes. Combined with reporting.spawnAndMonitor(), Java processes can be monitored during load tests.
* Add 'header-code' statistic which counts number of responses with different values for a given header. For instance, this can be used to graph cache misses/hits from Squid responses using the X-Cache header.

Bug Fixes:

* config: Add 'nodeload/config' module for configuring global parameters
* multiloop: polling time for next change in load or user profiles was always 1 second
* stats: Fix one-off error in Histogram.percentile wouldn't return the greatest number if it is greater than the number of buckets (i.e. in extra[]). Fix Uniques.clear() to actually reset count.
* nl.js: #issue/5: nl.js discarded URL query string and hash

## v0.2.0 (2010/12/01) ##

This release is a substantial, non-backwards-compatible rewrite of nodeload. The major features are:

* [npm](http://npmjs.org/) compatibility
* Independently usable modules: loop, stats, monitoring, http, reporting, and remote
* Addition of load and user profiles

Specific changes to note are:

* npm should be used to build the source

        [~/nodeload]> curl http://npmjs.org/install.sh | sh     # install npm if not already installed
        [~/nodeload]> npm link

* `nodeload` is renamed to `nl` and `nodeloadlib` to `nodeload`.

* addTest() / addRamp() / runTest() is replaced by run():

        var nl = require('nodeload');
        var loadtest = nl.run({ ... test specications ... }, ...);

* remoteTest() / remoteStart() is replaced by LoadTestCluster.run:

        var nl = require('nodeload');
        var cluster = new nl.LoadTestCluster(master:port, [slaves:port, ...]);
        cluster.run({ ... test specifications ...});

* Callbacks and most of the globals (except `HTTP_SERVER` and `REPORT_MANAGER`) have been removed. Instead EventEmitters are used throughout. For example, run() returns an instance of LoadTest, which emits 'update' and 'end' events, replacing the need for both `TEST_MONITOR` and the startTests() callback parameter.

* Scheduler has been replaced by MultiLoop, which also understands load & concurrency profiles.

* Statistics tracking works through event handlers now rather than by wrapping the loop function. See monitoring/monitor.js.

## v0.100.0 (2010/10/06) ##

This release adds nodeloadlib and moves to Dygraph for charting.

## v0.1.0 to v0.1.2 (2010/02/27) ##

Initial releases of nodeload. Tags correspond to node compatible versions. To find a version of node that's compatible with a tag release do `git show <tagname>`.

    For example: git show v0.1.1


nl.js
#!/usr/bin/env node
/*
 Copyright (c) 2010 Benjamin Schmaus
 Copyright (c) 2010 Jonathan Lee 

 Permission is hereby granted, free of charge, to any person
 obtaining a copy of this software and associated documentation
 files (the "Software"), to deal in the Software without
 restriction, including without limitation the rights to use,
 copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the
 Software is furnished to do so, subject to the following
 conditions:

 The above copyright notice and this permission notice shall be
 included in all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 OTHER DEALINGS IN THE SOFTWARE.
*/

/*jslint sub:true */
/*globals __dirname */

require.paths.unshift(__dirname);

var options = require('./lib/nl/options');
options.process();

if (!options.get('url')) {
    options.help();
}

var nl = require('./nodeload')
            .quiet()
            .setMonitorIntervalMs(options.get('reportInterval') * 1000);

function puts(text) { if (!options.get('quiet')) { console.log(text); } }
function pad(str, width) { return str + (new Array(width-str.length)).join(' '); }
function printItem(name, val, padLength) {
    if (padLength === undefined) { padLength = 40; }
    puts(pad(name + ':', padLength) + ' ' + val);
}

var testStart;
var host = options.get('host');
var test = nl.run({
    name: host,
    host: options.get('host'),
    port: options.get('port'),
    requestGenerator: options.get('requestGenerator'),
    method: options.get('method'),
    path: options.get('path'),
    requestData: options.get('requestData'),
    numUsers: options.get('numClients'),
    numRequests: options.get('numRequests'),
    timeLimit: options.get('timeLimit'),
    targetRps: options.get('targetRps'),
    stats: ['latency', 'result-codes', 'request-bytes', 'response-bytes']
});

test.on('start', function(tests) { testStart = new Date(); });
test.on('update', function(interval, stats) {
    puts(pad('Completed ' +stats[host]['result-codes'].length+ ' requests', 40));
});
test.on('end', function() {

    var stats = test.stats[host];
    var elapsedSeconds = ((new Date()) - testStart)/1000;

    puts('');
    printItem('Server', options.get('host') + ':' + options.get('port'));

    if (options.get('requestGeneratorModule') === undefined) {
        printItem('HTTP Method', options.get('method'));
        printItem('Document Path', options.get('path'));
    } else {
        printItem('Request Generator', options.get('requestGeneratorModule'));
    }

    printItem('Concurrency Level', options.get('numClients'));
    printItem('Number of requests', stats['result-codes'].length);
    printItem('Body bytes transferred', stats['request-bytes'].total + stats['response-bytes'].total);
    printItem('Elapsed time (s)', elapsedSeconds.toFixed(2));
    printItem('Requests per second', (stats['result-codes'].length/elapsedSeconds).toFixed(2));
    printItem('Mean time per request (ms)', stats['latency'].mean().toFixed(2));
    printItem('Time per request standard deviation', stats['latency'].stddev().toFixed(2));
    
    puts('\nPercentages of requests served within a certain time (ms)');
    printItem('  Min', stats['latency'].min, 6);
    printItem('  Avg', stats['latency'].mean().toFixed(1), 6);
    printItem('  50%', stats['latency'].percentile(0.5), 6);
    printItem('  95%', stats['latency'].percentile(0.95), 6);
    printItem('  99%', stats['latency'].percentile(0.99), 6);
    printItem('  Max', stats['latency'].max, 6);

    process.exit(0);
});
test.start();